File: Backend/pipeline/publish_telemetry.py
#!/usr/bin/env python3
"""
publish_telemetry.py
--------------------
Caterpillar Telemetry MQTT Publisher Simulator in Python.
Reads telemetry.csv and publishes records in real-time.
"""

import os
import sys
import time
import json
import csv
from urllib.parse import urlparse
import paho.mqtt.client as mqtt

# Configurations
CSV_PATH = os.path.join(os.path.dirname(__file__), 'telemetry.csv')
BROKER_URL_RAW = os.getenv('MQTT_BROKER_URL', 'mqtt://broker.hivemq.com:1883')
PUBLISH_INTERVAL_MS = int(os.getenv('PUBLISH_INTERVAL_MS', '1000')) # default 1 sec
CLIENT_ID = f"cat_publisher_py_{int(time.time())}"
BASE_TOPIC = 'caterpillar/telemetry'

def parse_broker_url(url_str):
    """
    Parses a broker URL into host and port.
    Supports mqtt://, tcp://, and raw hosts.
    """
    url = urlparse(url_str)
    host = url.hostname or url_str
    if ':' in host:
        host, port_str = host.split(':', 1)
        port = int(port_str)
    else:
        port = url.port or 1883
    return host, port

def main():
    print('----------------------------------------------------')
    print('Caterpillar Telemetry MQTT Publisher Simulator (Python)')
    print(f'CSV File: {CSV_PATH}')
    
    host, port = parse_broker_url(BROKER_URL_RAW)
    print(f'MQTT Broker: {host}:{port}')
    print(f'Publish Interval: {PUBLISH_INTERVAL_MS}ms per row')
    print('----------------------------------------------------')

    if not os.path.exists(CSV_PATH):
        print(f"Error: CSV File not found at {CSV_PATH}", file=sys.stderr)
        sys.exit(1)

    # Load CSV data into memory
    queue = []
    try:
        with open(CSV_PATH, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                # Basic schema alignment and type casting for json serialization
                payload = {
                    "timestamp": row.get("timestamp"),
                    "equipmentId": row.get("equipmentId"),
                    "equipmentType": row.get("equipmentType", "Excavator"),
                    "dealerId": row.get("dealerId") or None,
                    "siteId": row.get("siteId") or None,
                    "operatorId": row.get("operatorId") or None,
                    "engineStatus": row.get("engineStatus", "ON"),
                    "fuelLevel": float(row.get("fuelLevel", 0)) if row.get("fuelLevel") else 0.0,
                    "engineHours": float(row.get("engineHours", 0)) if row.get("engineHours") else 0.0,
                    "idleHours": float(row.get("idleHours", 0)) if row.get("idleHours") else 0.0,
                    "speed": float(row.get("speed", 0)) if row.get("speed") else 0.0,
                    "latitude": float(row.get("latitude")) if row.get("latitude") else None,
                    "longitude": float(row.get("longitude")) if row.get("longitude") else None,
                    "engineTemperature": float(row.get("engineTemperature", 0)) if row.get("engineTemperature") else 0.0,
                    "hydraulicPressure": float(row.get("hydraulicPressure", 0)) if row.get("hydraulicPressure") else 0.0,
                    "batteryVoltage": float(row.get("batteryVoltage", 0)) if row.get("batteryVoltage") else 0.0,
                    "loadPercentage": float(row.get("loadPercentage", 0)) if row.get("loadPercentage") else 0.0,
                    "vibrationLevel": float(row.get("vibrationLevel", 0)) if row.get("vibrationLevel") else 0.0,
                    "rentalStatus": row.get("rentalStatus", "Working")
                }
                queue.append(payload)
    except Exception as e:
        print(f"Error reading CSV: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Loaded {len(queue)} telemetry rows into transmission queue.")

    # Connect to broker
    client = mqtt.Client(client_id=CLIENT_ID, clean_session=True)
    
    connected = False
    def on_connect(c, userdata, flags, rc):
        nonlocal connected
        if rc == 0:
            print("✓ Successfully connected to MQTT Broker!")
            connected = True
        else:
            print(f"MQTT connection failed with code {rc}")

    client.on_connect = on_connect

    try:
        client.connect(host, port, keepalive=60)
        client.loop_start()
    except Exception as e:
        print(f"Could not connect to MQTT broker: {e}")
        print("Running in DRY-RUN mode...")
    
    print("Beginning transmission loop. Press Ctrl+C to terminate...")
    currentIndex = 0

    try:
        while True:
            if not queue:
                print("Queue is empty, exiting.")
                break

            payload = queue[currentIndex]
            topic = f"{BASE_TOPIC}/{payload['equipmentType']}/{payload['equipmentId']}"
            payload_str = json.dumps(payload)

            if connected:
                res = client.publish(topic, payload_str, qos=0)
                res.wait_for_publish()
                # Print progress
                if currentIndex % 5 == 0 or currentIndex == 0:
                    print(f"[MQTT] Published | Topic: {topic} | Speed: {payload['speed']} km/h | Temp: {payload['engineTemperature']}°C")
            else:
                # Dry run
                if currentIndex % 2 == 0 or currentIndex == 0:
                    print(f"[DRY-RUN] Packet #{currentIndex} | Topic: {topic}")
                    print(json.dumps(payload, indent=2))
                    print('------------------------------------------------')

            currentIndex = (currentIndex + 1) % len(queue)
            time.sleep(PUBLISH_INTERVAL_MS / 1000.0)

    except KeyboardInterrupt:
        print("\nPublisher terminated by user.")
    finally:
        client.loop_stop()
        client.disconnect()

if __name__ == '__main__':
    main()

File: Backend/pipeline/publish_telemetry.js
const fs = require('fs');
const path = require('path');
const readline = require('readline');

// Import mqtt safely
let mqtt;
try {
  mqtt = require('mqtt');
} catch (e) {
  console.error('\n[WARNING] The "mqtt" package is not installed.');
  console.error('Please run "npm install mqtt" in your terminal to enable publishing.');
  console.error('Running in DRY-RUN mode (JSON payloads will be printed to console only).\n');
}

// Configurations
const CSV_PATH = path.join(__dirname, 'telemetry.csv');
const BROKER_URL = process.env.MQTT_BROKER_URL || 'mqtt://broker.hivemq.com:1883';
const CLIENT_ID = `cat_publisher_${Math.random().toString(16).substring(2, 8)}`;
const BASE_TOPIC = 'caterpillar/telemetry';

// Speed configuration: publish a row every X milliseconds
const PUBLISH_INTERVAL_MS = parseInt(process.env.PUBLISH_INTERVAL_MS) || 100; // 10 rows per second default

// CSV Columns index map (headers from CSV file)
const HEADERS = [
  'Timestamp', 'EquipmentID', 'EquipmentType', 'DealerID', 'SiteID', 'OperatorID',
  'Latitude', 'Longitude', 'EngineStatus', 'Speed', 'EngineHours', 'IdleHours',
  'FuelLevel', 'FuelConsumed', 'HydraulicPressure', 'EngineTemperature',
  'BatteryVoltage', 'VibrationLevel', 'LoadPercentage', 'RentalStatus'
];

console.log('----------------------------------------------------');
console.log('Caterpillar Telemetry MQTT Publisher Simulator');
console.log(`CSV File: ${CSV_PATH}`);
console.log(`MQTT Broker: ${mqtt ? BROKER_URL : 'None (Dry-run)'}`);
console.log(`Publish Interval: ${PUBLISH_INTERVAL_MS}ms per row`);
console.log('----------------------------------------------------');

// Connect to broker if mqtt is available
let client = null;
if (mqtt) {
  console.log('Connecting to MQTT Broker...');
  client = mqtt.connect(BROKER_URL, {
    clientId: CLIENT_ID,
    clean: true,
    connectTimeout: 5000,
    reconnectPeriod: 1000
  });

  client.on('connect', () => {
    console.log('✓ Successfully connected to MQTT Broker!');
    startPublishing();
  });

  client.on('error', (err) => {
    console.error('MQTT connection error:', err.message);
    console.log('Switching to DRY-RUN mode...');
    client = null;
    startPublishing();
  });
} else {
  // Start in dry-run mode immediately
  setTimeout(startPublishing, 1000);
}

function startPublishing() {
  if (!fs.existsSync(CSV_PATH)) {
    console.error(`Error: CSV File not found at ${CSV_PATH}`);
    console.error('Please generate telemetry.csv first by running "node scripts/generate_data.js"');
    process.exit(1);
  }

  const fileStream = fs.createReadStream(CSV_PATH);
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  });

  let isHeader = true;
  const queue = [];
  
  rl.on('line', (line) => {
    if (isHeader) {
      isHeader = false; // Skip CSV headers row
      return;
    }
    if (line.trim()) {
      queue.push(line);
    }
  });

  rl.on('close', () => {
    console.log(`Loaded ${queue.length.toLocaleString()} telemetry rows into transmission queue.`);
    console.log('Beginning transmission loop. Press Ctrl+C to terminate...');
    
    let currentIndex = 0;
    
    function sendNext() {
      if (currentIndex >= queue.length) {
        console.log('Reached the end of telemetry.csv. Restarting loop...');
        currentIndex = 0;
      }
      
      const line = queue[currentIndex];
      currentIndex++;
      
      // Parse CSV line
      const values = line.split(',');
      if (values.length < HEADERS.length) {
        // Skip malformed lines
        setTimeout(sendNext, 0);
        return;
      }
      
      // Map columns to structured JSON object
      const telemetryObj = {
        timestamp: values[0],
        equipmentId: values[1],
        equipmentType: values[2],
        dealerId: values[3],
        siteId: values[4] === 'NULL' ? null : values[4],
        operatorId: values[5] === 'NULL' ? null : values[5],
        gps: {
          latitude: values[6] ? parseFloat(values[6]) : null,
          longitude: values[7] ? parseFloat(values[7]) : null
        },
        engine: {
          status: values[8],
          hours: parseFloat(values[10]),
          idleHours: parseFloat(values[11]),
          load: parseInt(values[18]),
          temperature: parseInt(values[15])
        },
        speed: parseInt(values[9]),
        fuel: {
          level: parseFloat(values[12]),
          consumed: parseFloat(values[13])
        },
        diagnostics: {
          hydraulicPressure: parseInt(values[14]),
          batteryVoltage: parseFloat(values[16]),
          vibrationLevel: parseFloat(values[17])
        },
        rentalStatus: values[19]
      };
      
      const payloadString = JSON.stringify(telemetryObj);
      const topic = `${BASE_TOPIC}/${telemetryObj.equipmentType}/${telemetryObj.equipmentId}`;
      
      if (client && client.connected) {
        client.publish(topic, payloadString, { qos: 0 }, (err) => {
          if (err) {
            console.error('Failed to publish packet:', err);
          }
        });
        
        // Print progress every 100 rows to keep console output clean
        if (currentIndex % 100 === 0 || currentIndex === 1) {
          console.log(`[MQTT] Published ${currentIndex.toLocaleString()} rows. Current: ${topic} -> Speed: ${telemetryObj.speed} km/h, Temp: ${telemetryObj.engine.temperature}°C`);
        }
      } else {
        // Dry run console logger (limited to print every 50 steps to avoid scroll lock)
        if (currentIndex % 50 === 0 || currentIndex === 1) {
          console.log(`[DRY-RUN] Packet #${currentIndex} | Topic: ${topic}`);
          console.log(JSON.stringify(telemetryObj, null, 2));
          console.log('------------------------------------------------');
        }
      }
      
      setTimeout(sendNext, PUBLISH_INTERVAL_MS);
    }
    
    // Start loop
    sendNext();
  });
}
